extensions/parquet/table/src/main/java/io/deephaven/parquet/table/layout/ParquetFlatPartitionedLayout.java

```diff
@@ -48,20 +63,33 @@ public String toString() {
     }
 
     @Override
-    public synchronized void findKeys(@NotNull final Consumer<ParquetTableLocationKey> locationKeyObserver) {
-        try (final DirectoryStream<Path> parquetFileStream =
-                Files.newDirectoryStream(tableRootDirectory.toPath(), ParquetUtils::fileNameMatches)) {
-            for (final Path parquetFilePath : parquetFileStream) {
-                ParquetTableLocationKey locationKey = cache.get(parquetFilePath);
-                if (locationKey == null) {
-                    locationKey = locationKey(parquetFilePath, readInstructions);
-                    if (!locationKey.verifyFileReader()) {
-                        continue;
+    public void findKeys(@NotNull final Consumer<ParquetTableLocationKey> locationKeyObserver) {
+        final SeekableChannelsProvider provider = SeekableChannelsProviderLoader.getInstance().fromServiceLoader(
+                tableRootDirectory, readInstructions.getSpecialInstructions());
+        final boolean isFileURI = FILE_URI_SCHEME.equals(tableRootDirectory.getScheme());
+        try (final Stream<URI> stream = provider.list(tableRootDirectory)) {
+            final Predicate<URI> uriFilter;
+            if (isFileURI) {
+                uriFilter = uri -> {
+                    final String filename = new File(uri).getName();
+                    return filename.endsWith(ParquetUtils.PARQUET_FILE_EXTENSION) && filename.charAt(0) != '.';
+                };
+            } else {
+                uriFilter = uri -> uri.getPath().endsWith(ParquetUtils.PARQUET_FILE_EXTENSION);
+            }
+            stream.filter(uriFilter).forEach(uri -> {
+                synchronized (ParquetFlatPartitionedLayout.this) {
```

Comment:
You could consider using `computeifAbsent` with a `Collections.synchronizedMap`.